### PR TITLE
Abilities finetuning & left of map in ORs

### LINKED ISSUE #1201: Show button not obvious - might need improvement
Created an 1846 multi player game to try to trigger issue 1130.
When doing that, I let PRR buy Lake Shore Line.
When I do Lake Shore Line appear as an aibility in the game.

But when I click on Show for that ability Steamboat Company appear....

I would have guessed that it should show the Lake Shore Line ability there...

{
  "id": 23,
  "description": "1130 take 2",
  "user": {
    "id": 5,
    "name": "a"
  },
  "players": [
    {
      "id": 5,
      "name": "a"
    },
    {
      "id": 9,
      "name": "a1@a.a"
    },
    {
      "id": 6,
      "name": "b"
    },
    {
      "id": 10,
      "name": "b1"
    },
    {
      "id": 7,
      "name": "c"
    }
  ],
  "max_players": 5,
  "title": "1846",
  "settings": {
    "seed": 2.3038643112949798e+38
  },
  "status": "active",
  "turn": 1,
  "round": "Draft Round",
  "acting": [
    7
  ],
  "result": {},
  "actions": [
    {
      "type": "bid",
      "entity": "c",
      "entity_type": "player",
      "id": 1,
      "company": "O&I",
      "price": 40
    },
    {
      "type": "bid",
      "entity": "b1",
      "entity_type": "player",
      "id": 2,
      "company": "MPC",
      "price": 60
    },
    {
      "type": "bid",
      "entity": "b",
      "entity_type": "player",
      "id": 3,
      "company": "LSL",
      "price": 40
    },
    {
      "type": "bid",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 4,
      "company": "BIG4",
      "price": 40
    },
    {
      "type": "bid",
      "entity": "a",
      "entity_type": "player",
      "id": 5,
      "company": "TBC",
      "price": 60
    },
    {
      "type": "bid",
      "entity": "c",
      "entity_type": "player",
      "id": 6,
      "company": "C&WI",
      "price": 60
    },
    {
      "type": "bid",
      "entity": "b1",
      "entity_type": "player",
      "id": 7,
      "company": "MAIL",
      "price": 80
    },
    {
      "type": "bid",
      "entity": "b",
      "entity_type": "player",
      "id": 8,
      "company": "SC",
      "price": 40
    },
    {
      "type": "bid",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 9,
      "company": "MC",
      "price": 40
    },
    {
      "type": "bid",
      "entity": "a",
      "entity_type": "player",
      "id": 10,
      "company": "MS",
      "price": 60
    },
    {
      "type": "par",
      "entity": "a",
      "entity_type": "player",
      "id": 11,
      "corporation": "PRR",
      "share_price": "40,0,4"
    },
    {
      "type": "par",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 12,
      "corporation": "NYC",
      "share_price": "40,0,4"
    },
    {
      "type": "par",
      "entity": "b",
      "entity_type": "player",
      "id": 13,
      "corporation": "B&O",
      "share_price": "40,0,4"
    },
    {
      "type": "par",
      "entity": "b1",
      "entity_type": "player",
      "id": 14,
      "corporation": "C&O",
      "share_price": "40,0,4"
    },
    {
      "type": "par",
      "entity": "c",
      "entity_type": "player",
      "id": 15,
      "corporation": "ERIE",
      "share_price": "40,0,4"
    },
    {
      "type": "pass",
      "entity": "a",
      "entity_type": "player",
      "id": 16
    },
    {
      "type": "pass",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 17
    },
    {
      "type": "pass",
      "entity": "b",
      "entity_type": "player",
      "id": 18
    },
    {
      "type": "pass",
      "entity": "b1",
      "entity_type": "player",
      "id": 19
    },
    {
      "type": "pass",
      "entity": "c",
      "entity_type": "player",
      "id": 20
    },
    {
      "type": "undo",
      "entity": "MS",
      "entity_type": "minor",
      "id": 21
    },
    {
      "type": "undo",
      "entity": "c",
      "entity_type": "player",
      "id": 22
    },
    {
      "type": "undo",
      "entity": "b1",
      "entity_type": "player",
      "id": 23
    },
    {
      "type": "undo",
      "entity": "b",
      "entity_type": "player",
      "id": 24
    },
    {
      "type": "undo",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 25
    },
    {
      "type": "buy_shares",
      "entity": "a",
      "entity_type": "player",
      "id": 26,
      "shares": [
        "PRR_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 27,
      "shares": [
        "NYC_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "b",
      "entity_type": "player",
      "id": 28,
      "shares": [
        "B&O_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "b1",
      "entity_type": "player",
      "id": 29,
      "shares": [
        "C&O_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "c",
      "entity_type": "player",
      "id": 30,
      "shares": [
        "ERIE_1"
      ]
    },
    {
      "type": "undo",
      "entity": "a",
      "entity_type": "player",
      "id": 31
    },
    {
      "type": "undo",
      "entity": "c",
      "entity_type": "player",
      "id": 32
    },
    {
      "type": "undo",
      "entity": "b1",
      "entity_type": "player",
      "id": 33
    },
    {
      "type": "undo",
      "entity": "b",
      "entity_type": "player",
      "id": 34
    },
    {
      "type": "undo",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 35
    },
    {
      "type": "undo",
      "entity": "a",
      "entity_type": "player",
      "id": 36
    },
    {
      "type": "undo",
      "entity": "c",
      "entity_type": "player",
      "id": 37
    },
    {
      "type": "undo",
      "entity": "b1",
      "entity_type": "player",
      "id": 38
    },
    {
      "type": "undo",
      "entity": "b",
      "entity_type": "player",
      "id": 39
    },
    {
      "type": "undo",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 40
    },
    {
      "type": "par",
      "entity": "a",
      "entity_type": "player",
      "id": 41,
      "corporation": "PRR",
      "share_price": "60,0,6"
    },
    {
      "type": "par",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 42,
      "corporation": "NYC",
      "share_price": "80,0,8"
    },
    {
      "type": "par",
      "entity": "b",
      "entity_type": "player",
      "id": 43,
      "corporation": "B&O",
      "share_price": "90,0,9"
    },
    {
      "type": "par",
      "entity": "b1",
      "entity_type": "player",
      "id": 44,
      "corporation": "C&O",
      "share_price": "80,0,8"
    },
    {
      "type": "par",
      "entity": "c",
      "entity_type": "player",
      "id": 45,
      "corporation": "ERIE",
      "share_price": "100,0,10"
    },
    {
      "type": "buy_shares",
      "entity": "a",
      "entity_type": "player",
      "id": 46,
      "shares": [
        "PRR_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 47,
      "shares": [
        "NYC_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "b",
      "entity_type": "player",
      "id": 48,
      "shares": [
        "B&O_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "b1",
      "entity_type": "player",
      "id": 49,
      "shares": [
        "C&O_1"
      ]
    },
    {
      "type": "buy_shares",
      "entity": "c",
      "entity_type": "player",
      "id": 50,
      "shares": [
        "ERIE_1"
      ]
    },
    {
      "type": "pass",
      "entity": "a",
      "entity_type": "player",
      "id": 51
    },
    {
      "type": "pass",
      "entity": "a1@a.a",
      "entity_type": "player",
      "id": 52
    },
    {
      "type": "pass",
      "entity": "b",
      "entity_type": "player",
      "id": 53
    },
    {
      "type": "pass",
      "entity": "b1",
      "entity_type": "player",
      "id": 54
    },
    {
      "type": "pass",
      "entity": "c",
      "entity_type": "player",
      "id": 55
    },
    {
      "type": "lay_tile",
      "entity": "MS",
      "entity_type": "minor",
      "id": 56,
      "hex": "C13",
      "tile": "8-0",
      "rotation": 4
    },
    {
      "type": "lay_tile",
      "entity": "MS",
      "entity_type": "minor",
      "id": 57,
      "hex": "D12",
      "tile": "9-0",
      "rotation": 0
    },
    {
      "type": "lay_tile",
      "entity": "BIG4",
      "entity_type": "minor",
      "id": 58,
      "hex": "G9",
      "tile": "57-0",
      "rotation": 1
    },
    {
      "type": "lay_tile",
      "entity": "BIG4",
      "entity_type": "minor",
      "id": 59,
      "hex": "G7",
      "tile": "57-1",
      "rotation": 1
    },
    {
      "type": "run_routes",
      "entity": "BIG4",
      "entity_type": "minor",
      "id": 60,
      "routes": [
        {
          "train": "2-1",
          "connections": [
            [
              "G9",
              "G7"
            ]
          ]
        }
      ]
    },
    {
      "type": "pass",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 61
    },
    {
      "type": "lay_tile",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 62,
      "hex": "E19",
      "tile": "8-1",
      "rotation": 5
    },
    {
      "type": "lay_tile",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 63,
      "hex": "E17",
      "tile": "293-0",
      "rotation": 1
    },
    {
      "type": "pass",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 64
    },
    {
      "type": "buy_train",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 65,
      "train": "2-2",
      "price": 80
    },
    {
      "type": "pass",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 66
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 67
    },
    {
      "type": "pass",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 68
    },
    {
      "type": "pass",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 69
    },
    {
      "type": "lay_tile",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 70,
      "hex": "D18",
      "tile": "8-2",
      "rotation": 4
    },
    {
      "type": "pass",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 71
    },
    {
      "type": "buy_company",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 72,
      "company": "LSL",
      "price": 40
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 73
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 74
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 75
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 76
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 77
    },
    {
      "type": "buy_company",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 78,
      "company": "LSL",
      "price": 20
    },
    {
      "type": "undo",
      "entity": "NYC",
      "entity_type": "corporation",
      "id": 79
    },
    {
      "type": "undo",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 80
    },
    {
      "type": "undo",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 81
    },
    {
      "type": "buy_company",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 82,
      "company": "LSL",
      "price": 20
    },
    {
      "type": "place_token",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 83,
      "city": "293-0-0",
      "slot": 0
    },
    {
      "type": "undo",
      "entity": "PRR",
      "entity_type": "corporation",
      "id": 84
    }
  ],
  "loaded": true,
  "created_at": 1595848447,
  "updated_at": 1595848579
}

### PR DESCRIPTION
**Pull request by yzemaze**
- Abilities in h3 to fit with other action headers
- Abilities in ORs on top for consistency. I’d move them below actions or even below corp, but that’s your decision.
- Description of company rendered right below the name in ORs (before: below last company).
- Truncation on names improved to cut at least 4 chars (=> no "Mitsubishi Ferr…" etc.) and show more chars of company names. Imposed limit should work fine with 20rem width, current company names and font. For names with many wide chars (CAPS, m, w etc.) this may become a problem later on, but then it’s an easy change.

might resolve #1201

### CODE REVIEW

File: assets/app/view/game/abilities.rb [hunk 1]
```diff
@@ -8,17 +8,22 @@ class Abilities < Snabberb::Component
       needs :selected_company, default: nil, store: true
       needs :show_other_abilities, default: false, store: true
 
-      ABILITIES = %i[tile_lay teleport assign_hexes assign_corporation token exchange].freeze
-
       def render
         companies = @game.companies.select { |company| !company.closed? && actions_for(company).any? }
         return h(:div) if companies.empty? || @game.round.current_entity.company?
 
+        op_round = @game.round.is_a?(Engine::Round::Operating)
```
Comment (tobymao) [inline assets/app/view/game/abilities.rb:15]: let's pass this in as a need instead, show_desciption true or false

File: assets/app/view/game/abilities.rb [hunk 2]
```diff
@@ -8,17 +8,22 @@ class Abilities < Snabberb::Component
...
         current, others = companies.partition { |company| @game.current_entity.owner == company.owner.owner }
 
+        props = {
+          style: {
+            margin: '0 1rem 0 0',
+            display: op_round && current.any? ? '' : 'inline-block',
```
Comment (tobymao) [inline assets/app/view/game/abilities.rb:21]: props[:style] = 'inlinle-block' if...

File: assets/app/view/game/abilities.rb [hunk 3]
```diff
@@ -29,39 +34,64 @@ def render
...
 
-        if companies.include?(@selected_company)
-          children << h(:div, { style: { margin: '0.5rem 0' } }, @selected_company.desc)
-          children.concat(render_actions)
-        end
+        children << render_selected_company_desc if !op_round && companies.include?(@selected_company)
 
-        h(:div, children)
+        props = { style: { marginBottom: '0.5rem' } }
+        h(:div, props, children)
```
Comment (tobymao) [inline assets/app/view/game/abilities.rb:51]: inline

File: assets/app/view/game/abilities.rb [hunk 4]
```diff
@@ -29,39 +34,64 @@ def render
...
+          }
+          children << h('button.button', props, @show_other_abilities ? 'Hide Others' : 'Show Others')
+          children.concat(render_companies(others, op_round)) if @show_other_abilities
         end
 
-        if companies.include?(@selected_company)
-          children << h(:div, { style: { margin: '0.5rem 0' } }, @selected_company.desc)
-          children.concat(render_actions)
-        end
+        children << render_selected_company_desc if !op_round && companies.include?(@selected_company)
```
Comment (tobymao) [inline assets/app/view/game/abilities.rb:48]: hm this looks incorrect, what if you want to exchange a share in the operating round?

the name of this method is bad, it's not just company description, it's also the actions

File: assets/app/view/game/abilities.rb [hunk 5]
```diff
@@ -29,39 +34,64 @@ def render
...
-          company_name = company_name[0..15] + '...' if company_name.size > 15
-
           owner_name = company.owner.id
-          owner_name = owner_name[0..10] + '...' if owner_name.size > 10
+          c_size = company_name.size
+          o_size = owner_name.size
+          limit = 36
+          if c_size + o_size > limit
+            if o_size < 5
+              company_name = company_name[0..(limit - 7)] + '…' if c_size > (limit - 3)
```
Comment (tobymao) [inline assets/app/view/game/abilities.rb:86]: let's make a truncation helper module

File: assets/app/view/game/round/operating.rb [hunk 6]
```diff
@@ -22,7 +22,7 @@ def render
           @current_actions = round.actions_for(entity)
           entity = entity.owner if entity.company?
 
-          left = [h(UndoAndPass, pass: @current_actions.include?('pass'))]
+          left = [h(Game::Abilities, user: @user, game: @game), h(UndoAndPass, pass: @current_actions.include?('pass'))]
```
Comment (tobymao) [inline assets/app/view/game/round/operating.rb:25]: multi lines 
left = [ 
h(),
h(),
]

File: assets/app/view/game_card.rb [hunk 7]
```diff
@@ -146,7 +146,7 @@ def render_body
       }
 
       p_elm = players.map.with_index do |player, index|
-        short_name = player['name'].length > 19 ? player['name'][0...18] + '…' : player['name']
+        short_name = player['name'].length > 19 ? player['name'][0..15] + '…' : player['name']
```
Comment (tobymao) [inline assets/app/view/game_card.rb:149]: truncation library

File: assets/app/view/game/abilities.rb (hunk 4 above)
Comment (yzemaze) [reply assets/app/view/game/abilities.rb:48]: > hm this looks incorrect, what if you want to exchange a share in the operating round?

Dougo works fine:
image
image

Comment (tobymao) [reply assets/app/view/game/abilities.rb:48]: hm how does it work then? i must not understand the code

Comment (yzemaze) [reply assets/app/view/game/abilities.rb:48]: Maybe. It’s a bit tricky. Abilities rendering is triggered at several places, depending if op_round or not.

Comment (yzemaze) [reply assets/app/view/game/abilities.rb:48]: render_companies does render abilities if op_round
L48 does it if !op_round

**Comment by yzemaze** [comment]
finetuning and truncation => separate PRs
move to left of map => scrapped
A bigger approach will be necessary down the line anyway and then it likely won’t be beneficiary to have different behaviours of abilities depending on current round.

**Pull request closed without merging**